classes/security/Validation.php

```diff
@@ -474,6 +482,83 @@ public static function canAdminister($administeredUserId, $administratorUserId)
         // There were no conflicting roles. Permit administration.
         return true;
     }
+
+    /**
+     * Get the user's administration level
+     *
+     * @param int   $administeredUserId     User ID of user to potentially administer
+     * @param int   $administratorUserId    User ID of user who wants to do the administrating
+     * @param int   $contextId              The journal/context Id
+     * 
+     * @return int The authorized adminstration level 
+     */
+    public static function getAdministrationLevel(int $administeredUserId, int $administratorUserId, int $contextId = null): int
+    {
+        // You can administer yourself
+        if ($administeredUserId == $administratorUserId) {
+            return self::ADMINISTRATION_FULL;
+        }
+
+        $filteredSiteAdminUserGroups = Repo::userGroup()
+            ->getCollector()
+            ->filterByContextIds([\PKP\core\PKPApplication::CONTEXT_SITE])
+            ->filterByRoleIds([Role::ROLE_ID_SITE_ADMIN]);
+
+        // You cannot adminster administrators
+        if ($filteredSiteAdminUserGroups->filterByUserIds([$administeredUserId])->getCount() > 0) {
+            return self::ADMINISTRATION_PROHIBITED;
+        }
+
+        // Otherwise, administrators can administer everyone
+        if ($filteredSiteAdminUserGroups->filterByUserIds([$administratorUserId])->getCount() > 0) {
+            return self::ADMINISTRATION_FULL;
+        }
+
+        // Check for administered user group assignments in other contexts
+        // that the administrator user doesn't have a manager role in.
+        $userGroupsCount = Repo::userGroup()
```

Comment:
I suspect this won't perform well on large databases as it executes queries in a loop.

What we need is the number of user groups that the `$administeredUserId` is in, but that `$administratorUserId` doesn't manage. I would suggest getting a list of IDs for all user groups that `$administratorUserId` manages, and a list of all user groups that `$administeredUserId` participates in. If the second list intersected the first list contains any items, then full administration should be prevented. That would just require two queries. (We could rewrite it down to a single query, but that would require adding some single-use functions to the `Collector` class that I don't think would be justified.)

Reply:
@asmecher you are right and I should have considered to refactor it . But the implementation that I did is bit different than what you proposed . If I understand correctly , all we need to determine any context that has the `administered user group` assigned but the `administrator user` don't have nay `manager` role in that context. So basically we can do a diff between the both of the user case context list . And thats sounds logical and easier. But correct me if I understand it wrong .

Reply:
You're right -- it should be a diff, not an intersect!